src/main/java/org/sonatype/nexus/ci/nxrm/v2/ComponentUploaderNxrm2.groovy

```diff
@@ -0,0 +1,93 @@
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package org.sonatype.nexus.ci.nxrm.v2
+
+import com.sonatype.nexus.api.exception.RepositoryManagerException
+import com.sonatype.nexus.api.repository.v2.GAV
+import com.sonatype.nexus.api.repository.v2.MavenAsset
+import com.sonatype.nexus.api.repository.v2.RepositoryManagerV2Client
+
+import org.sonatype.nexus.ci.config.Nxrm2Configuration
+import org.sonatype.nexus.ci.nxrm.ComponentUploader
+import org.sonatype.nexus.ci.nxrm.MavenCoordinate
+
+import groovy.transform.InheritConstructors
+import hudson.FilePath
+import hudson.model.Result
+
+import static org.sonatype.nexus.ci.util.RepositoryManagerClientUtil.nexus2Client
+
+@SuppressWarnings(['CatchException', 'AbcMetric', 'MethodSize'])
+@InheritConstructors
+class ComponentUploaderNxrm2
+    extends ComponentUploader
+{
+  @Override
+  void upload(final Map<MavenCoordinate, List<RemoteMavenAsset>> remoteMavenComponents,
+                        String nxrmRepositoryId)
+  {
+    def nxrmClient = getRepositoryManagerClient(nxrmConfiguration)
+
+    remoteMavenComponents.each { mavenCoordinate, remoteMavenAssets ->
+      logger.println("Uploading Maven asset with groupId: ${mavenCoordinate.groupId} " +
+          "artifactId: ${mavenCoordinate.artifactId} version: ${mavenCoordinate.version} " +
+          "To repository: ${nxrmRepositoryId}")
+
+      remoteMavenAssets.each { remoteMavenAsset ->
+        try {
+          def localFile = File.createTempFile(remoteMavenAsset.RemotePath.getName(), '.tmp')
+          remoteMavenAsset.RemotePath.copyTo(new FilePath(localFile))
+
+          def gav = new GAV(envVars.expand(mavenCoordinate.groupId), envVars.expand(mavenCoordinate.artifactId),
+              envVars.expand(mavenCoordinate.version), envVars.expand(mavenCoordinate.packaging))
+
+          def mavenFile = new MavenAsset(localFile, envVars.expand(remoteMavenAsset.Asset.extension),
+              envVars.expand(remoteMavenAsset.Asset.classifier))
+
+          try {
+            nxrmClient.uploadComponent(nxrmRepositoryId, gav, [mavenFile])
+          }
+          catch (RepositoryManagerException ex) {
+            throw new IOException(ex)
+          }
+          finally {
+            localFile.delete()
+          }
+        }
+        catch (IOException ex) {
+          final String uploadFailed = "Upload of ${remoteMavenAsset.Asset.filePath} failed"
+
+          logger.println(uploadFailed)
+          logger.println('Failing build due to failure to upload file to Nexus Repository Manager Publisher')
+          run.setResult(Result.FAILURE)
+          throw new IOException(uploadFailed, ex)
+        }
+      }
+
+      logger.println('Successfully Uploaded Maven Assets')
+    }
+  }
+
+  RepositoryManagerV2Client getRepositoryManagerClient(final Nxrm2Configuration nexusConfiguration) {
```

Comment:
I think if this method would take `NxrmConfiguration` instead it would eliminate a compiler warning where it is called above.